### PR TITLE
Added runtime metrics and sampling rules to startup logs

### PR DESCRIPTION
**Pull request by bojbrook**
## Description
Updated startup logs to include runtime metrics and sampling rules configuration

## Checklist
- [ ] Added to the correct milestone.
- [x ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Library documentation is updated.
- [ ] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).

### CODE REVIEW

File: ddtrace/internal/debug.py [hunk 1]
```diff
@@ -123,12 +123,14 @@ def collect(tracer):
         tracer_enabled=tracer.enabled,
         sampler_type=type(tracer.sampler).__name__ if tracer.sampler else "N/A",
         priority_sampler_type=type(tracer.priority_sampler).__name__ if tracer.priority_sampler else "N/A",
+        sampler_rules=ddtrace.config.sampler_rules,
```
Comment (brettlangdon) [inline ddtrace/internal/debug.py:126]: The issue with using a config setting like this is that if they manually configure a sampler without the env variables then we will not get the proper rules.

Instead we might want to consider something like:

```python
sampler_rules = None

if isinstance(tracer.sampler, DatadogSampler):
    sampler_rules = [str(rule) for rule in tracer.sampler.rules]

return dict(
    ...
    sampler_rules=sampler_rules,
)
```


We could also consider doing the same with the `DatadogSampler.limiter`

```python
sampler_limiter = None
if isinstance(tracer.sampler, DatadogSampler):
    sampler_limiter = str(tracer.sampler.limiter)

return dict(
    ...
    sampler_limiter=sampler_limiter,
)
```


... or... we could consider add a `__str__`/`__repr__` onto each of our sampler classes so we can just do:

```python
return dict(
    ...
    sampler=str(tracer.sampler),
)
```

To get something like `sampler="DatadogSampler(limiter=RateLimiter(rate_limit=100, ...), rules=[SamplingRule(sample_rate=1.0, service="my-srvc", ...), ...])"`.

Going this route could make it easier than basically unpacking all the metadata from a given sampler.

What do you think?

File: ddtrace/internal/debug.py [hunk 2]
```diff
@@ -123,12 +123,14 @@ def collect(tracer):
...
         service=ddtrace.config.service or "",
         debug=ddtrace.tracer.log.isEnabledFor(logging.DEBUG),
         enabled_cli="ddtrace" in os.getenv("PYTHONPATH", ""),
         analytics_enabled=ddtrace.config.analytics_enabled,
         log_injection_enabled=ddtrace.config.logs_injection,
         health_metrics_enabled=ddtrace.config.health_metrics_enabled,
+        runtime_metrics_enabled=ddtrace.config.runtime_metrics,
```
[the remark this answers is not in the document]
Comment (bojbrook) [reply ddtrace/internal/debug.py:133]: That's a good point. Though by grabbing the RuntimeWorkers enabled attribute this creates a circular dependency that I haven't been able to solve yet.
```
from ddtrace.internal.runtime.runtime_metrics import RuntimeWorker
...
runtime_metrics_enabled=RuntimeWorker.enabled,
...
````
Resulting in 
```
File "/Users/bowen.brooks/Documents/APM_Sandbox/dd-trace-py/ddtrace/internal/runtime/runtime_metrics.py", line 76, in RuntimeWorker
    tracer = attr.ib(type=ddtrace.tracer, default=None)
AttributeError: partially initialized module 'ddtrace' has no attribute 'tracer' (most likely due to a circular import)
```

File: ddtrace/settings/config.py [hunk 3]
```diff
@@ -147,6 +147,10 @@ def __init__(self):
         # Raise certain errors only if in testing raise mode to prevent crashing in production with non-critical errors
         self._raise = asbool(os.getenv("DD_TESTING_RAISE", False))
 
+        self.runtime_metrics = asbool(get_env("runtime", "metrics_enabled", default=False))
+
+        self.sampler_rules = parse_tags_str(get_env("trace", "sampling_rules", default=""))
```
[the remark this answers is not in the document]
Comment (bojbrook) [reply ddtrace/settings/config.py:152]: I'll remove these options from the config file 👍

File: ddtrace/internal/debug.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (brettlangdon) [reply ddtrace/internal/debug.py:133]: the only idea I have for this is, do the import in `def collect()` rather that at the top of the file, that way importing is deferred until after the tracer module is fully loaded.


not sure I love this, but only idea I have for the moment.

File: ddtrace/internal/debug.py (hunk 1 above)
Comment (bojbrook) [reply ddtrace/internal/debug.py]: I went with you first suggestion but if you feel that the second route is cleaner I'm happy to implement it

File: releasenotes/notes/add-runtime-and-sampling-rules-to-startup-logs-18a8a8236d4131f5.yaml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+---
+upgrade:
+  - |
+    Added runtime metrics enabled and sampling rules to startup logs
```
Comment (brettlangdon) [inline releasenotes/notes/add-runtime-and-sampling-rules-to-startup-logs-18a8a8236d4131f5.yaml:4]:
```suggestion
    Added runtime metrics status and sampling rules to startup logs.
```

File: releasenotes/notes/add-runtime-and-sampling-rules-to-startup-logs-18a8a8236d4131f5.yaml [hunk 5]
```diff
@@ -0,0 +1,4 @@
+---
+upgrade:
```
Comment (brettlangdon) [inline releasenotes/notes/add-runtime-and-sampling-rules-to-startup-logs-18a8a8236d4131f5.yaml:2]:
```suggestion
features:
```


or maybe

```suggestion
other:
```

"upgrade" is like "in order to upgrade you should make X change"

File: tests/integration/test_debug.py [hunk 6]
```diff
@@ -264,6 +267,32 @@ def test_tracer_info_level_log(self):
...
+    monkeypatch.setenv("DD_RUNTIME_METRICS_ENABLED", "true")
+    _, _, status, _ = ddtrace_run_python_code_in_subprocess(
+        """
+import ddtrace
+from ddtrace.internal import debug
+f = debug.collect(ddtrace.tracer)
+assert f.get("runtime_metrics_enabled") is True
+""",
+    )
+    assert status == 0
```
Comment (brettlangdon) [inline tests/integration/test_debug.py:293]: Not a strong opinion to make this change, but figured also testing the default with no env set might be good.

```suggestion
    # default, no env variable set
    _, _, status, _ = ddtrace_run_python_code_in_subprocess(
        """
import ddtrace
from ddtrace.internal import debug
f = debug.collect(ddtrace.tracer)
assert f.get("runtime_metrics_enabled") is False
""",
    )
    assert status == 0

    # Explicitly set env variable
    monkeypatch.setenv("DD_RUNTIME_METRICS_ENABLED", "true")
    _, _, status, _ = ddtrace_run_python_code_in_subprocess(
        """
import ddtrace
from ddtrace.internal import debug
f = debug.collect(ddtrace.tracer)
assert f.get("runtime_metrics_enabled") is True
""",
    )
    assert status == 0
```

Comment (brettlangdon) [reply tests/integration/test_debug.py:293]: This kind of happens with `test_standard_tags` already, however the difference with this suggestion is also testing the default with `ddtrace-run` vs just importing `ddtrace`

**Review by brettlangdon — COMMENTED** [review]
small nits/suggestions, otherwise lgtm

File: tests/integration/test_debug.py [hunk 7]
```diff
@@ -264,6 +267,44 @@ def test_tracer_info_level_log(self):
         assert tracer.log.log.mock_calls == []
 
 
+def test_runtime_metrics_enabled_via_manual_start():
```
Comment (brettlangdon) [inline tests/integration/test_debug.py:270]: I wonder if this should be a subprocess test to handle any cleanup from enabling runtime metrics 🤔

**Review by brettlangdon — DISMISSED** [review]
minor thought on one test, otherwise lgtm

File: tests/integration/test_debug.py (hunk 7 above)
Comment (bojbrook) [reply tests/integration/test_debug.py:270]: I'll make this change

**Review by brettlangdon — APPROVED** [review]

**Review by Kyle-Verhoog — APPROVED** [review]
👍 👍 nice - great job @bojbrook!

**Pull request merged**